streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -706,13 +662,17 @@ void runOnce() {
                     totalProcessed += processed;
                 }
 
+                log.debug("TaskManager#process handled {} records; invoking TaskManager#punctuate", processed);
+
                 final int punctuated = taskManager.punctuate();
                 final long punctuateLatency = advanceNowAndComputeLatency();
                 totalPunctuateLatency += punctuateLatency;
                 if (punctuated > 0) {
                     punctuateSensor.record(punctuateLatency / (double) punctuated, now);
                 }
 
+                log.debug("TaskManager#punctuate executed: {}", punctuated);
```

Comment:
trace?

Reply:
Note, this isn't saying that we invoked a single punctuator, but that we invoked _all_ the punctuators that are runnable right now. It's also a top-level phase of executing StreamThread, so I don't think we can leave it out of the debug logs without telling an incomplete story of what's happening. If we log this (or any of the proposed logs) at trace level, the only consequence is that users who want to _debug_ StreamThread would have to use "trace" level instead of "debug" level. This seems to add unnecessary complexity to the logging.

IMO, it's better to reserve "trace" level for _very_ low-level logs, such as logging the progress of individual documents through the processors, and use "debug" level for higher-level summary logs like the ones in this PR.